targets/psa_switch/externs/psa_checksum.cpp

```diff
@@ -0,0 +1,186 @@
+/* Copyright 2021 SYRMIA LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Dusan Krdzic (dusan.krdzic@syrmia.com)
+ *
+ */
+
+
+#include "psa_checksum.h"
+#include <bm/bm_sim/calculations.h>
+
+namespace {
+
+static const unsigned char hex_digits[] = {'0','1','2','3','4','5','6','7','8','9','A','B','C','D','E','F'};
+
+uint8_t
+convert_hex_str_to_u8(std::string hex) {
+  int a;
+  std::istringstream(hex) >> std::hex >> a;
+  return static_cast<uint8_t>(a);
+}
+
+std::string
+restore_packet_value_from_input_fields(const std::vector<bm::Field> &fields) {
+  std::string hex_ascii;
+  std::string hex_hex;
+  const uint8_t base = 16;
+  uint8_t num_hex_digits = 0;
+  bm::Data input;
+  input.set(0);
+  uint16_t current_bits_offset = 0;
+
+  // Vector fields has separate field values as its elements.
+  // They need to be concatenated to one single data which represents value from packet.
+  for(int i = fields.size() - 1 ; i >= 0 ; i--) {
+    bm::Data shift_value;
+    shift_value.shift_left(bm::Data(fields.at(i).get<uint64_t>()), current_bits_offset);
+    input.add(input, shift_value);
+    current_bits_offset += fields.at(i).get_nbits();
+  }
+
+  // Data input is in decimal format and it needs to be converted to hexadecimal,
+  // because value in packet is in hexadecimal.
+  while (input != 0) {
+    hex_ascii += hex_digits[input % base];
+    num_hex_digits++;
+    input.divide(input, base);
+  }
+  std::reverse(hex_ascii.begin(), hex_ascii.end());
+
+  while(num_hex_digits * 4 < current_bits_offset) {
+    hex_ascii.insert(hex_ascii.begin(), 1, '0');
+    num_hex_digits++;
+  }
+
+  // Hash algorithms process the data in chunks of bytes (chars).
+  // If the hex value from input is "4F", it semantically means 0x4F.
+  // If "4F" is directly passed to hash algorithm, it will process it as 0x34 0x46.
+  // So, it is important to convert "4F" to char with value 0x4F in ASCII table,
+  // which is 'O'. Then the hash algorithm will process it as 0x4F.
+  for (size_t i = 0; i < hex_ascii.size(); i += 2) {
+    std::string hex_byte;
+    hex_byte += hex_ascii[i];
+    if(i != hex_ascii.size() - 1)
+      hex_byte += hex_ascii[i + 1];
+    uint8_t num = convert_hex_str_to_u8(hex_byte);
+    hex_hex += num;
+  }
```

Comment:
Hash algorithms process the data in chunks of bytes (chars). If the hex value from input is "4F", it semantically means 0x4F. If "4F" is directly passed to hash algorithm, it will process it as 0x34 0x46. So, it is important to convert "4F" to char with value 0x4F in ASCII table, which is 'O'. Then the hash algorithm will process it as 0x4F.